### PR TITLE
Revamp sub/3 to resolve most issues with gsub (and sub with "g")

### LINKED ISSUE #1206: gsub with empty regex doesn't terminate
Calling `gsub` with an empty regex will run forever. Example:

``` console
$ jq -n '"" | gsub(""; "")'
```

Not sure if this is a bug or expected behavior, but terminating with an error might be more appropriate :)

### LINKED ISSUE #1425: gsub  breaks \b for checking word boundaries
When I run the following program with `jq -n`:

```
"hello world"|"\\b(?<i>[[:lower:]])" as $rx|gsub($rx;.i|ascii_upcase)
```

Expected Output:

> `"Hello World"`

Actual Output:

> `"HELLO WORLD"`

I've got a workaround for my particular use case:

```
"hello world"|"\\b(?<fst>[[:lower:]])(?<rst>[[:lower:]]*)\\b" as $rx|gsub($rx;"\(.fst|ascii_upcase)\(.rst)")
```

### LINKED ISSUE #1572: faster version of _nwise
Here are plug-in replacements for the existing defs.  These replacements use 0-arity filters to effect the recursion, and are ridiculously faster even for arrays of size 10^4. (*) 

```
def _nwise($n):
  def nw: if length <= $n then . else .[0:$n] , (.[$n:] | nw) end;
  nw;

def _nwise(a; $n): a | _nwise($n);

```

(*) [[range(0;n)] | _nwise(3)] | length
```
            builtin.jq   replacement
n           u+s          u+s
10,000      23.0s        0.027s
```

### LINKED ISSUE #2354: gsub lookahead cannot allocate memory
**Describe the bug**
When using regex lookahead `jq` freezes, eats up all system memory, prints an error message and eventually aborts

**To Reproduce**

    $ jq -n '"qux" | gsub("(?<=u)"; "u")'
    "quux"

    $ free -g
                  total        used        free      shared  buff/cache   available
    Mem:             15           5           6           0           3           8
    Swap:             1           0           1

    $ jq -n '"qux" | gsub("(?=u)"; "u")'
    error: cannot allocate memory
    [1]    123456 abort (core dumped)  jq -n '"qux" | gsub("(?=u)";  #"u")'

**Expected behavior**
`jq` outputs the result and exit

**Environment**
 - Ubuntu 20.04.2 LTS
 - jq-1.6

**Additional context**
I'm aware that this may be an issue in Oniguruma but I wasn't able to reproduce it due to my lack of experience with C

### LINKED ISSUE #2532: gsub not works while or lookahead and normal pattern
**Describe the bug**

I want output capture pattern if matched else empty
example input
```
not a cd cmd
cd /ok
```

I want extract `/ok` which follows `cd `, and replace invalid lines to empty, expected
```
""
"/ok"
```

my try is `jq -nR 'inputs|gsub("^(?!cd ).*$|^cd "; "")'`
`^(?!cd ).*$` works, `^cd ` also works, but failed while out them together.

**To Reproduce**
```

[root build]#(echo "not a cd cmd"; echo "cd /ok") | ./jq -nR 'inputs|gsub("^cd "; "")'
"not a cd cmd"
"/ok"
[root build]#(echo "not a cd cmd"; echo "cd /ok") | ./jq -nR 'inputs|gsub("^(?!cd ).*$"; "")'
""
"cd /ok"
[root build]#(echo "not a cd cmd"; echo "cd /ok") | ./jq -nR 'inputs|gsub("^(?!cd ).*$|^cd "; "")'
""
""

```

**Expected behavior**

jq works likes perl
```
[root build]#(echo "not a cd cmd"; echo "cd /ok")
not a cd cmd
cd /ok
[root build]#(echo "not a cd cmd"; echo "cd /ok") | perl -pe 's/^(?!cd ).*$|^cd //g'

/ok
[root build]#
```

**Environment (please complete the following information):**
 - OS: CentOS 7
 - jq version: jq-1.6

**Additional context**

### PR DESCRIPTION
**Pull request by pkoppstein**
The primary purpose of this commit (which supercedes PR https://github.com/jqlang/jq/pull/2624) is to rectify most problems with `gsub` (and also `sub` with the "g" option), in particular fix #1425 ('\b'), fix #2354 (lookahead), and fix #2532 (regex == "^(?!cd ).*$|^cd ";"")).

This commit also partly resolves #2148 and also resolves #1206 in that `gsub` no longer loops infinitely; however, because the new `gsub` depends critically on match(_;"g"), the behavior when regex == "" is sometimes non-standard. [*1]

Since the new sub/3 relies on uniq/1, that has been added as well [*2].

The documentation has been updated to reflect the fact that `sub` and `gsub` are intended to be regular in the second argument. [*3]

Also, _nwise/1 has been tweaked to take advantage of TCO.

Footnotes:

[*1] Using the new gsub, '"a" | gsub( ""; "a")' emits "aa" rather than "aaa" as would be standard.  This is nevertheless better than the infinite loop behavior of jq 1.6 in this case.

With one exception (as explained in [*2]), the new gsub is implemented as though match/2 behavior is correct.  That is, bugs in `gsub` behavior will most likely have their origin in `match/2`.

[*2] `uniq/1` adopts the Unix/Linux name and semantics; it is needed for the following test case:

gsub("(?=u)"; "u")
"qux"
"quux"

Without this functionality:

Test #23: 'gsub("(?=u)"; "u")' at line number 100
*** Expected "quux", but got "quuux" for test at line number 102: gsub("(?=u)"; "u")

The root of the problem here is `match`: if `match` is fixed, then gsub would not need `untie`.

The addition of `uniq` as a top-level function should be a non-issue relative to general concern about builtins.jq bloat: the line count of the new builtin.jq is significantly reduced overall, and the number of defs is actually reduced by 1 (from 111 (ignoring a redundant def) to 110).

[*3] See e.g. https://github.com/jqlang/jq/issues/513#issuecomment-50834811

### CODE REVIEW

**Comment by pkoppstein** [comment]
@itchyny  - Please note that the failing checks almost certainly have nothing to do with the changes in this PR.

Please also note that in the case of "no match" examples such as `"p" | gsub("q"; "a","b")`, this PR currently follows gojq, but that it would only require a small change to one line to produce what I would argue is actually the "correct" behavior (in the sense of regularity in the second position).

**Comment by pkoppstein** [comment]
@itchyny - as expected, the failing tests had nothing to do with this PR.  Since “all checks have passed”, it would be a good time to pull, before merging becomes very complex, wouldn’t it?

File: docs/content/manual/manual.yml
Comment (itchyny) [inline docs/content/manual/manual.yml]: This example is invalid and somehow not included in man.test. There is no `sub/1` and `output` should always be an array. Since you added examples, would you rebuild man.test to check if the CI passes?

File: docs/content/manual/manual.yml
Comment (itchyny) [inline docs/content/manual/manual.yml]: I'm still disagree with this naming, I'm very worried about this confuses the users. Consistency within the builtin functions is much important than external command name conventions. We need to remember that `uniq` is a stream oriented and `unique` is an array oriented. There're already filters having both array and stream versions like `first` so that's good.

File: docs/content/manual/manual.yml
Comment (itchyny) [inline docs/content/manual/manual.yml:2511]: Adding an example will help users understand this sentence.

File: tests/onig.test
Comment (itchyny) [inline tests/onig.test:78]: Nitpick but inconsistent spacing.

File: src/builtin.jq
Comment (itchyny) [inline src/builtin.jq]: Following patch isn't enough to fix the duplicate bug?
```diff
diff --git a/src/builtin.c b/src/builtin.c
index 9b2d9a2..03ab4d5 100644
--- a/src/builtin.c
+++ b/src/builtin.c
@@ -930,7 +930,7 @@ static jv f_match(jq_state *jq, jv input, jv regex, jv modifiers, jv testmode) {
         match = jv_object_set(match, jv_string("string"), jv_string(""));
         match = jv_object_set(match, jv_string("captures"), jv_array());
         result = jv_array_append(result, match);
-        start += 1;
+        start = (const UChar*)(input_string+region->end[0]+1);
         continue;
       }
 
```
```sh
 $ ./jq -nc '"qqqqqqux" | match("(?=u)"; "g")'
{"offset":6,"length":0,"string":"","captures":[]}
```

File: src/builtin.jq
Comment (itchyny) [inline src/builtin.jq:103]: We can use omitting `else .` by flipping the condition.
```jq
def _nwise($n): def _nwise: if length > $n then .[:$n], (.[$n:] | _nwise) end; _nwise;
```

File: src/builtin.jq
Comment (itchyny) [inline src/builtin.jq]: Inconsistent tab indentation.

File: src/builtin.jq
Comment (itchyny) [inline src/builtin.jq:132]: Removing `// $in` passes the tests, but this is necessary. Would you add a test that doesn't match?

**Review by itchyny — CHANGES_REQUESTED** [review]

File: src/builtin.jq
Comment (itchyny) [reply src/builtin.jq:132]: ↑ I thought this but this is actually unnecessary!

File: src/builtin.jq
Comment (pkoppstein) [reply src/builtin.jq:103]: Being able to omit `else` is not in jq 1.6, so my reason for sticking with the 1.6 syntax is that it should be easy for someone familiar with 1.6 to understand code that might well be "copied and pasted" into a jq program, or indeed a jaq program.

File: docs/content/manual/manual.yml
Comment (pkoppstein) [reply docs/content/manual/manual.yml]: As per notes elsewhere, this example is very old, and was always wrong. It has been replaced.

File: docs/content/manual/manual.yml
Comment (pkoppstein) [reply docs/content/manual/manual.yml:2511]: Example added in https://github.com/jqlang/jq/pull/2641/commits/5373191e64d9763c14da46b1657828779b74f462

File: src/builtin.jq
Comment (pkoppstein) [reply src/builtin.jq]: Hmmm.  This seems to me to be a case of apples and oranges.  The first `reduce` occurs in the form `[reduce ... ] as $_`.  Anyway, in my view, it's fine to put a `reduce` (or an `if..then..else..end`) on one line if its fits.  But feel free to tweak :-)

File: src/builtin.jq
Comment (pkoppstein) [reply src/builtin.jq:132]: Please see detailed notes elsewhere regarding the three variants of this line. Assuming you want gsub fixed in the next release, I'd suggest merging and then tweaking, otherwise this "branch" will fall too far behind and be quite difficult to manage later as there are so many files involved, two or three of which are the subject of many commits.

File: src/builtin.jq
Comment (itchyny) [reply src/builtin.jq]: I just noticed you're mixing space and tab indentations...

Comment (pkoppstein) [reply src/builtin.jq]: Sorry about that. Two tabs in builtin.jq removed in  https://github.com/pkoppstein/jq/commit/3015ec8d8ddc6ae9780d0881f87896437edee440

File: src/builtin.c
Comment (itchyny) [inline src/builtin.c]: There is tab indentation here but we prefer spaces.

**Review by itchyny — APPROVED** [review]
LGTM!

File: src/builtin.c
Comment (pkoppstein) [reply src/builtin.c]: Fixed (in 832734b5cdc49935a4a8530c8c492fd908da40b2); also fixed the editor setting.

**Comment by pkoppstein** [comment]
@itchyny wrote:
> LGTM!

Thanks.  I manually forced  the restarting of the checking process, and now there's a sign "All checks have passed" again.  But github also tells me:

> You’re not [authorized](https://docs.github.com/articles/about-protected-branches/) to merge this pull request

**Comment by itchyny** [comment]
@pkoppstein Can you kick the ci again (by rebasing again?) to make sure you can merge?

**Comment by pkoppstein** [comment]
@itchyny - As expected, my git/github skills are not up to the task.

In fact, I may have made things worse.  Using GitHub Desktop, I clicked on "Update from upstream/master"
which resulted in two new commits that I didn't intend. 

However, GitHub tells me now that "This branch [gsub] is not behind the upstream jqlang:master".
That looks promising, doesn't it?

**Comment by itchyny** [comment]
@pkoppstein That's good. However , if you want to drop the merge commit, you can use `git reset --hard HEAD^` and `git rebase upstream/master`.

**Comment by pkoppstein** [comment]
@itchyny wrote:
> That's good.

Apparently, good enough for you to merge into jqlang:  

image

I'll try to avoid messing things up in the meantime.

If something needs adjusting, please feel free to make changes: "Maintainers are allowed to edit this pull request."

**Comment by itchyny** [comment]
If you don't mind, I'll merge this PR. Would you rephrase the PR description (which contains the outdated descriptions about uniq)? Note that `fix #XXX` links the issue toe be closed by this PR so the that's important. Thanks.

**Comment by pkoppstein** [comment]
@itchyny wrote:

> If you don't mind, I'll merge this PR

Since I don’t have the permissions to do so, that would be great. I think it would be best if you made any adjustments when doing so. I will undoubtedly just make a mess of things. Tx.

**Comment by emanuele6** [comment]
I think there is a bug in this implementation: If you use the `"g"` flag, and the pattern matches up to the end of the string, it will replace both that match and the empty string at the end of the string if the pattern is able to match the empty string:

```jq
"1" | gsub("1*"; "2")
# "22"
```

EDIT: actually, that is a bug with `_match_impl/3`

**Comment by pkoppstein** [comment]
> I think there is a bug in this implementation ... "1" | gsub("1*"; "2")

Interestingly, 7 of the 8 regex engines at regex101.com indicate that "1" ~ /1*/ twice, evidently because
/1*/ eats the 1, and then matches the EOS as well.  I wonder why it's Rust that's the odd one out here.

Perhaps the the moral of the story is that users should avoid using "g" in cases like this.

**Pull request merged**
